Kernel/FlatTerm.hpp

```diff
@@ -55,38 +55,56 @@ class FlatTerm
   struct Entry
   {
     Entry() = default;
-    Entry(EntryTag tag, unsigned num) { _info.tag=tag; _info.number=num; }
-    Entry(Term* ptr) : _ptr(ptr) { ASS_EQ(tag(), FUN_TERM_PTR); }
-
-    inline EntryTag tag() const { return static_cast<EntryTag>(_info.tag); }
-    inline unsigned number() const { return _info.number; }
-    inline Term* ptr() const { return _ptr; }
-    inline bool isVar() const { return tag()==VAR; }
-    inline bool isVar(unsigned num) const { return isVar() && number()==num; }
-    inline bool isFun() const { return tag()==FUN || tag()==FUN_UNEXPANDED; }
-    inline bool isFun(unsigned num) const { return isFun() && number()==num; }
+    Entry(EntryTag tag, unsigned num) { _setTag(tag); _setNumber(num); }
+    Entry(Term* term) { _setTerm(term); ASS_EQ(_tag(), FUN_TERM_PTR); }
+
+    inline bool isVar() const { return _tag()==VAR; }
+    inline bool isVar(unsigned num) const { return isVar() && _number()==num; }
+    inline bool isFun() const { return _tag()==FUN || _tag()==FUN_UNEXPANDED; }
+    inline bool isFun(unsigned num) const { return isFun() && _number()==num; }
     /**
      * Should be called when @b isFun() is true.
      * If @b tag()==FUN_UNEXPANDED, it fills out entries for the functions
      * arguments with FUN_UNEXPANDED values. Otherwise does nothing.
      */
     void expand();
 
-    union {
-      Term* _ptr;
-      struct {
-	unsigned tag : 4;
-	unsigned number : 28;
-      } _info;
-    };
+    uint64_t _content;
+
+    static constexpr unsigned
+      TAG_BITS_START = 0,
+      TAG_BITS_END = TAG_BITS_START + 3,
+      NUMBER_BITS_START = TAG_BITS_END,
+      NUMBER_BITS_END = NUMBER_BITS_START + 27,
```

Comment:
You can have up to 64 - 3 = 61 bits for `number` here, if you want. 27 is a bit random, I might go for 29 (29 + 3 = 32), 32, or 61. Performance for any of these will be ballpark the same but you might shave a few cycles off here or there with one choice or another.